mlflow/utils/async_logging/async_logging_queue.py

```diff
@@ -0,0 +1,161 @@
+"""
+Defines an AsyncLoggingQueue that provides async fashion logging of metrics/tags/params using
+ queue based approach.
+"""
+
+import atexit
+import logging
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from queue import Empty, Queue
+
+from mlflow.entities.metric import Metric
+from mlflow.entities.param import Param
+from mlflow.entities.run_tag import RunTag
+from mlflow.utils.async_logging.run_batch import RunBatch
+from mlflow.utils.async_logging.run_operations import RunOperations
+
+_logger = logging.getLogger(__name__)
+
+
+class AsyncLoggingQueue:
+    """
+    This is a queue based run data processor that queues incoming batches and processes them using
+    single worker thread.
+    """
+
+    def __init__(self, logging_func: callable([str, [Metric], [Param], [RunTag]])) -> None:
+        """
+        Initializes an AsyncLoggingQueue object.
+
+        Args:
+            logging_func: A callable function that takes in four arguments: a string
+                representing the run_id, a list of Metric objects,
+                a list of Param objects, and a list of RunTag objects.
+        """
+        self._queue = Queue()  # Dict[str, Queue]
+        self._logging_func = logging_func
+        self._continue_to_log_data = threading.Event()
+        # Keeping max_workers=1 so that there are no two threads
+        self._BATCH_LOGGING_THREADPOOL = ThreadPoolExecutor(max_workers=1)
+
+        self._BATCH_STATUS_CHECK_THREADPOOL = ThreadPoolExecutor(max_workers=1)
```

Comment:
I don't think we need to limit the size of `self._BATCH_STATUS_CHECK_THREADPOOL` to `1`, these threads are just waiting for the finish signal. We can use a decent number like `500` or so

Reply:
Regarding this, we can increase the thread worker count, but since this is a queue, threads wont get to work in parallel and have to wait for prior batches to be cleared before current thread waiting gets unblocked by batch.completion_event.set() call. So in a way adding more threads does not seem to be useful to me.

Let me know what you think.

Reply:
There is overhead of collecting and reassigning worker if we have `max_workers=1`, that's why I suggest adding more workers. We can use a relatively small number, e.g., `5`/`10`.